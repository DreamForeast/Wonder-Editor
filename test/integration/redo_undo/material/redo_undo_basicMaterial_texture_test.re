open Wonder_jest;

open Expect;

open Expect.Operators;

open Sinon;

let _ =
  describe("redo_undo: basicMaterial texture", () => {
    let sandbox = getSandboxDefaultVal();
    let _getFromArray = (array, index) => ArrayService.getNth(index, array);
    beforeEach(() => {
      sandbox := createSandbox();
      MainEditorSceneTool.initStateAndGl(~sandbox, ());
      EventListenerTool.buildFakeDom()
      |> EventListenerTool.stubGetElementByIdReturnFakeDom;
    });
    afterEach(() => restoreSandbox(refJsObjToSandbox(sandbox^)));
    describe("test simulate set currentSceneTreeNode", () => {
      let _simulateTwiceDragTexture = assetTreeDomRecord => {
        let assetComponent = BuildComponentTool.buildAssetComponent();
        let firstTextureDomIndex =
          assetTreeDomRecord
          |> MainEditorAssetNodeTool.OperateTwoLayer.getFirstTextureDomIndex;
        let secondTextureDomIndex =
          assetTreeDomRecord
          |> MainEditorAssetNodeTool.OperateTwoLayer.getSecondTextureDomIndex;

        BaseEventTool.triggerComponentEvent(
          assetComponent,
          BasicMaterialEventTool.triggerFileDragStartEvent(
            firstTextureDomIndex,
          ),
        );
        let inspectorComponent =
          BuildComponentTool.buildInspectorComponent(
            TestTool.buildEmptyAppState(),
            InspectorTool.buildFakeAllShowComponentConfig(),
          );
        BaseEventTool.triggerComponentEvent(
          inspectorComponent,
          BasicMaterialEventTool.triggerTextureDragEnterEvent,
        );
        BaseEventTool.triggerComponentEvent(
          inspectorComponent,
          BasicMaterialEventTool.triggerTextureDragDropEvent,
        );
        let assetComponent2 = BuildComponentTool.buildAssetComponent();
        BaseEventTool.triggerComponentEvent(
          assetComponent2,
          BasicMaterialEventTool.triggerFileDragStartEvent(
            secondTextureDomIndex,
          ),
        );
        let inspectorComponent2 =
          BuildComponentTool.buildInspectorComponent(
            TestTool.buildEmptyAppState(),
            InspectorTool.buildFakeAllShowComponentConfig(),
          );
        BaseEventTool.triggerComponentEvent(
          inspectorComponent2,
          BasicMaterialEventTool.triggerTextureDragEnterEvent,
        );
        BaseEventTool.triggerComponentEvent(
          inspectorComponent2,
          BasicMaterialEventTool.triggerTextureDragDropEvent,
        );
      };
      beforeEach(() => {
        MainEditorSceneTool.createDefaultScene(
          sandbox,
          () => {
            MainEditorAssetTool.initAssetTree();
            MainEditorSceneTool.setFirstBoxTobeCurrentSceneTreeNode();
          },
        );
        CurrentSelectSourceEditorService.setCurrentSelectSource(
          EditorType.SceneTree,
        )
        |> StateLogicService.getAndSetEditorState;
      });
      afterEach(() =>
        StateAssetService.getState()
        |> CurrentNodeDataAssetService.clearCurrentNodeData
        |> CurrentNodeParentIdAssetService.clearCurrentNodeParentId
        |> StateAssetService.setState
        |> ignore
      );
      describe("test not remove texture", () => {
        describe("test undo operate", () => {
          test("test not undo", () => {
            let assetTreeDomRecord =
              MainEditorAssetTool.buildTwoLayerAssetTreeRoot();
            _simulateTwiceDragTexture(assetTreeDomRecord);

            BuildComponentTool.buildMaterialComponent(
              GameObjectTool.getCurrentGameObjectMaterial(),
            )
            |> ReactTestTool.createSnapshotAndMatch;
          });
          describe("test undo one step", () => {
            test("step which from second to first", () => {
              let assetTreeDomRecord =
                MainEditorAssetTool.buildTwoLayerAssetTreeRoot();
              _simulateTwiceDragTexture(assetTreeDomRecord);

              StateHistoryToolEditor.undo();

              BuildComponentTool.buildMaterialComponent(
                GameObjectTool.getCurrentGameObjectMaterial(),
              )
              |> ReactTestTool.createSnapshotAndMatch;
            });
            test("step which from second to zero", () => {
              let assetTreeDomRecord =
                MainEditorAssetTool.buildTwoLayerAssetTreeRoot();
              _simulateTwiceDragTexture(assetTreeDomRecord);

              StateHistoryToolEditor.undo();
              StateHistoryToolEditor.undo();

              BuildComponentTool.buildMaterialComponent(
                GameObjectTool.getCurrentGameObjectMaterial(),
              )
              |> ReactTestTool.createSnapshotAndMatch;
            });
          });
        });
        describe("test redo operate", () => {
          describe("test redo one step", () => {
            test("if not exec undo, redo one step, not change", () => {
              let assetTreeDomRecord =
                MainEditorAssetTool.buildTwoLayerAssetTreeRoot();
              _simulateTwiceDragTexture(assetTreeDomRecord);

              StateHistoryToolEditor.redo();

              BuildComponentTool.buildMaterialComponent(
                GameObjectTool.getCurrentGameObjectMaterial(),
              )
              |> ReactTestTool.createSnapshotAndMatch;
            });
            test(
              "undo step which from second to zero, redo step which from zero to first",
              () => {
              let assetTreeDomRecord =
                MainEditorAssetTool.buildTwoLayerAssetTreeRoot();
              _simulateTwiceDragTexture(assetTreeDomRecord);

              StateHistoryToolEditor.undo();
              StateHistoryToolEditor.undo();
              StateHistoryToolEditor.redo();

              BuildComponentTool.buildMaterialComponent(
                GameObjectTool.getCurrentGameObjectMaterial(),
              )
              |> ReactTestTool.createSnapshotAndMatch;
            });
          });
          describe("test redo two step", () =>
            test(
              "undo step which from second to zero, redo step which from zero to second",
              () => {
              let assetTreeDomRecord =
                MainEditorAssetTool.buildTwoLayerAssetTreeRoot();
              _simulateTwiceDragTexture(assetTreeDomRecord);

              StateHistoryToolEditor.undo();
              StateHistoryToolEditor.undo();
              StateHistoryToolEditor.redo();
              StateHistoryToolEditor.redo();

              BuildComponentTool.buildMaterialComponent(
                GameObjectTool.getCurrentGameObjectMaterial(),
              )
              |> ReactTestTool.createSnapshotAndMatch;
            })
          );
          describe("test redo three step", () =>
            test(
              "test if current step is last step, execute redo, not change", () => {
              let assetTreeDomRecord =
                MainEditorAssetTool.buildTwoLayerAssetTreeRoot();
              _simulateTwiceDragTexture(assetTreeDomRecord);

              StateHistoryToolEditor.undo();
              StateHistoryToolEditor.undo();
              StateHistoryToolEditor.redo();
              StateHistoryToolEditor.redo();
              StateHistoryToolEditor.redo();

              BuildComponentTool.buildMaterialComponent(
                GameObjectTool.getCurrentGameObjectMaterial(),
              )
              |> ReactTestTool.createSnapshotAndMatch;
            })
          );
        });
      });

      describe("test remove texture", () => {
        let _triggerRemoveTextureClickEvent = domChildren => {
          let textureArticle = _getFromArray(domChildren, 1);
          let button = _getFromArray(textureArticle##children, 3);
          BaseEventTool.triggerClickEvent(button);
        };

        describe("test undo operate", () => {
          test("test not undo", () => {
            let assetTreeDomRecord =
              MainEditorAssetTool.buildTwoLayerAssetTreeRoot();

            _simulateTwiceDragTexture(assetTreeDomRecord);
            BaseEventTool.triggerComponentEvent(
              BuildComponentTool.buildMaterialComponent(
                GameObjectTool.getCurrentGameObjectMaterial(),
              ),
              _triggerRemoveTextureClickEvent,
            );

            BuildComponentTool.buildMaterialComponent(
              GameObjectTool.getCurrentGameObjectMaterial(),
            )
            |> ReactTestTool.createSnapshotAndMatch;
          });
          describe("test undo one step", () =>
            test("step which from first to zero", () => {
              let assetTreeDomRecord =
                MainEditorAssetTool.buildTwoLayerAssetTreeRoot();
              _simulateTwiceDragTexture(assetTreeDomRecord);

              BaseEventTool.triggerComponentEvent(
                BuildComponentTool.buildMaterialComponent(
                  GameObjectTool.getCurrentGameObjectMaterial(),
                ),
                _triggerRemoveTextureClickEvent,
              );

              StateHistoryToolEditor.undo();

              BuildComponentTool.buildMaterialComponent(
                GameObjectTool.getCurrentGameObjectMaterial(),
              )
              |> ReactTestTool.createSnapshotAndMatch;
            })
          );
          describe("test redo operate", () =>
            describe("test redo one step", () => {
              test("if not exec undo, redo one step, not change", () => {
                let assetTreeDomRecord =
                  MainEditorAssetTool.buildTwoLayerAssetTreeRoot();
                _simulateTwiceDragTexture(assetTreeDomRecord);

                BaseEventTool.triggerComponentEvent(
                  BuildComponentTool.buildMaterialComponent(
                    GameObjectTool.getCurrentGameObjectMaterial(),
                  ),
                  _triggerRemoveTextureClickEvent,
                );

                StateHistoryToolEditor.redo();

                BuildComponentTool.buildMaterialComponent(
                  GameObjectTool.getCurrentGameObjectMaterial(),
                )
                |> ReactTestTool.createSnapshotAndMatch;
              });
              test(
                "undo step which from first to zero, redo step which from zero to first",
                () => {
                let assetTreeDomRecord =
                  MainEditorAssetTool.buildTwoLayerAssetTreeRoot();
                _simulateTwiceDragTexture(assetTreeDomRecord);

                BaseEventTool.triggerComponentEvent(
                  BuildComponentTool.buildMaterialComponent(
                    GameObjectTool.getCurrentGameObjectMaterial(),
                  ),
                  _triggerRemoveTextureClickEvent,
                );

                StateHistoryToolEditor.undo();
                StateHistoryToolEditor.redo();

                BuildComponentTool.buildMaterialComponent(
                  GameObjectTool.getCurrentGameObjectMaterial(),
                )
                |> ReactTestTool.createSnapshotAndMatch;
              });
            })
          );
        });
      });
    });
  });