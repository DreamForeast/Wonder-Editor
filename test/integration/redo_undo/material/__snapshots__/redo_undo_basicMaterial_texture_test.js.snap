// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`redo_undo: basicMaterial texture test simulate set currentSceneTreeNode test not remove texture test redo operate test redo one step if not exec undo, redo one step, not change 1`] = `
<article
  className="wonder-inspector-material"
>
  <article
    className="wonder-material-color"
  >
    <div
      className=""
    >
      <span
        className=""
      >
        color : 
      </span>
      <span
        className=""
      >
        #ffffff
      </span>
      <button
        className=""
        onClick={[Function]}
      >
        pick color
      </button>
    </div>
  </article>
  <article
    className="wonder-material-texture"
  >
    <div
      className="texture_ground"
      onDragEnter={[Function]}
      onDragLeave={[Function]}
      onDragOver={[Function]}
      onDrop={[Function]}
      style={
        Object {
          "opacity": "1",
        }
      }
    />
    <span
      className=""
    >
      texture:
    </span>
    <img
      src="texture5img"
    />
    <button
      className="texture_remove"
      onClick={[Function]}
    >
      remove
    </button>
  </article>
</article>
`;

exports[`redo_undo: basicMaterial texture test simulate set currentSceneTreeNode test not remove texture test redo operate test redo one step undo step which from second to zero, redo step which from zero to first 1`] = `
<article
  className="wonder-inspector-material"
>
  <article
    className="wonder-material-color"
  >
    <div
      className=""
    >
      <span
        className=""
      >
        color : 
      </span>
      <span
        className=""
      >
        #ffffff
      </span>
      <button
        className=""
        onClick={[Function]}
      >
        pick color
      </button>
    </div>
  </article>
  <article
    className="wonder-material-texture"
  >
    <div
      className="texture_ground"
      onDragEnter={[Function]}
      onDragLeave={[Function]}
      onDragOver={[Function]}
      onDrop={[Function]}
      style={
        Object {
          "opacity": "1",
        }
      }
    />
    <span
      className=""
    >
      texture:
    </span>
    <img
      src="texture3img"
    />
    <button
      className="texture_remove"
      onClick={[Function]}
    >
      remove
    </button>
  </article>
</article>
`;

exports[`redo_undo: basicMaterial texture test simulate set currentSceneTreeNode test not remove texture test redo operate test redo three step test if current step is last step, execute redo, not change 1`] = `
<article
  className="wonder-inspector-material"
>
  <article
    className="wonder-material-color"
  >
    <div
      className=""
    >
      <span
        className=""
      >
        color : 
      </span>
      <span
        className=""
      >
        #ffffff
      </span>
      <button
        className=""
        onClick={[Function]}
      >
        pick color
      </button>
    </div>
  </article>
  <article
    className="wonder-material-texture"
  >
    <div
      className="texture_ground"
      onDragEnter={[Function]}
      onDragLeave={[Function]}
      onDragOver={[Function]}
      onDrop={[Function]}
      style={
        Object {
          "opacity": "1",
        }
      }
    />
    <span
      className=""
    >
      texture:
    </span>
    <img
      src="texture5img"
    />
    <button
      className="texture_remove"
      onClick={[Function]}
    >
      remove
    </button>
  </article>
</article>
`;

exports[`redo_undo: basicMaterial texture test simulate set currentSceneTreeNode test not remove texture test redo operate test redo two step undo step which from second to zero, redo step which from zero to second 1`] = `
<article
  className="wonder-inspector-material"
>
  <article
    className="wonder-material-color"
  >
    <div
      className=""
    >
      <span
        className=""
      >
        color : 
      </span>
      <span
        className=""
      >
        #ffffff
      </span>
      <button
        className=""
        onClick={[Function]}
      >
        pick color
      </button>
    </div>
  </article>
  <article
    className="wonder-material-texture"
  >
    <div
      className="texture_ground"
      onDragEnter={[Function]}
      onDragLeave={[Function]}
      onDragOver={[Function]}
      onDrop={[Function]}
      style={
        Object {
          "opacity": "1",
        }
      }
    />
    <span
      className=""
    >
      texture:
    </span>
    <img
      src="texture5img"
    />
    <button
      className="texture_remove"
      onClick={[Function]}
    >
      remove
    </button>
  </article>
</article>
`;

exports[`redo_undo: basicMaterial texture test simulate set currentSceneTreeNode test not remove texture test undo operate test not undo 1`] = `
<article
  className="wonder-inspector-material"
>
  <article
    className="wonder-material-color"
  >
    <div
      className=""
    >
      <span
        className=""
      >
        color : 
      </span>
      <span
        className=""
      >
        #ffffff
      </span>
      <button
        className=""
        onClick={[Function]}
      >
        pick color
      </button>
    </div>
  </article>
  <article
    className="wonder-material-texture"
  >
    <div
      className="texture_ground"
      onDragEnter={[Function]}
      onDragLeave={[Function]}
      onDragOver={[Function]}
      onDrop={[Function]}
      style={
        Object {
          "opacity": "1",
        }
      }
    />
    <span
      className=""
    >
      texture:
    </span>
    <img
      src="texture5img"
    />
    <button
      className="texture_remove"
      onClick={[Function]}
    >
      remove
    </button>
  </article>
</article>
`;

exports[`redo_undo: basicMaterial texture test simulate set currentSceneTreeNode test not remove texture test undo operate test undo one step step which from second to first 1`] = `
<article
  className="wonder-inspector-material"
>
  <article
    className="wonder-material-color"
  >
    <div
      className=""
    >
      <span
        className=""
      >
        color : 
      </span>
      <span
        className=""
      >
        #ffffff
      </span>
      <button
        className=""
        onClick={[Function]}
      >
        pick color
      </button>
    </div>
  </article>
  <article
    className="wonder-material-texture"
  >
    <div
      className="texture_ground"
      onDragEnter={[Function]}
      onDragLeave={[Function]}
      onDragOver={[Function]}
      onDrop={[Function]}
      style={
        Object {
          "opacity": "1",
        }
      }
    />
    <span
      className=""
    >
      texture:
    </span>
    <img
      src="texture3img"
    />
    <button
      className="texture_remove"
      onClick={[Function]}
    >
      remove
    </button>
  </article>
</article>
`;

exports[`redo_undo: basicMaterial texture test simulate set currentSceneTreeNode test not remove texture test undo operate test undo one step step which from second to zero 1`] = `
<article
  className="wonder-inspector-material"
>
  <article
    className="wonder-material-color"
  >
    <div
      className=""
    >
      <span
        className=""
      >
        color : 
      </span>
      <span
        className=""
      >
        #ffffff
      </span>
      <button
        className=""
        onClick={[Function]}
      >
        pick color
      </button>
    </div>
  </article>
  <article
    className="wonder-material-texture"
  >
    <div
      className="texture_ground"
      onDragEnter={[Function]}
      onDragLeave={[Function]}
      onDragOver={[Function]}
      onDrop={[Function]}
      style={
        Object {
          "opacity": "1",
        }
      }
    />
    <span
      className=""
    >
      texture:
    </span>
    <img
      src="./public/img/null.jpg"
    />
    <button
      className="texture_remove"
      onClick={[Function]}
    >
      remove
    </button>
  </article>
</article>
`;

exports[`redo_undo: basicMaterial texture test simulate set currentSceneTreeNode test remove texture test undo operate test not undo 1`] = `
<article
  className="wonder-inspector-material"
>
  <article
    className="wonder-material-color"
  >
    <div
      className=""
    >
      <span
        className=""
      >
        color : 
      </span>
      <span
        className=""
      >
        #ffffff
      </span>
      <button
        className=""
        onClick={[Function]}
      >
        pick color
      </button>
    </div>
  </article>
  <article
    className="wonder-material-texture"
  >
    <div
      className="texture_ground"
      onDragEnter={[Function]}
      onDragLeave={[Function]}
      onDragOver={[Function]}
      onDrop={[Function]}
      style={
        Object {
          "opacity": "1",
        }
      }
    />
    <span
      className=""
    >
      texture:
    </span>
    <img
      src="./public/img/null.jpg"
    />
    <button
      className="texture_remove"
      onClick={[Function]}
    >
      remove
    </button>
  </article>
</article>
`;

exports[`redo_undo: basicMaterial texture test simulate set currentSceneTreeNode test remove texture test undo operate test redo operate test redo one step if not exec undo, redo one step, not change 1`] = `
<article
  className="wonder-inspector-material"
>
  <article
    className="wonder-material-color"
  >
    <div
      className=""
    >
      <span
        className=""
      >
        color : 
      </span>
      <span
        className=""
      >
        #ffffff
      </span>
      <button
        className=""
        onClick={[Function]}
      >
        pick color
      </button>
    </div>
  </article>
  <article
    className="wonder-material-texture"
  >
    <div
      className="texture_ground"
      onDragEnter={[Function]}
      onDragLeave={[Function]}
      onDragOver={[Function]}
      onDrop={[Function]}
      style={
        Object {
          "opacity": "1",
        }
      }
    />
    <span
      className=""
    >
      texture:
    </span>
    <img
      src="./public/img/null.jpg"
    />
    <button
      className="texture_remove"
      onClick={[Function]}
    >
      remove
    </button>
  </article>
</article>
`;

exports[`redo_undo: basicMaterial texture test simulate set currentSceneTreeNode test remove texture test undo operate test redo operate test redo one step undo step which from first to zero, redo step which from zero to first 1`] = `
<article
  className="wonder-inspector-material"
>
  <article
    className="wonder-material-color"
  >
    <div
      className=""
    >
      <span
        className=""
      >
        color : 
      </span>
      <span
        className=""
      >
        #ffffff
      </span>
      <button
        className=""
        onClick={[Function]}
      >
        pick color
      </button>
    </div>
  </article>
  <article
    className="wonder-material-texture"
  >
    <div
      className="texture_ground"
      onDragEnter={[Function]}
      onDragLeave={[Function]}
      onDragOver={[Function]}
      onDrop={[Function]}
      style={
        Object {
          "opacity": "1",
        }
      }
    />
    <span
      className=""
    >
      texture:
    </span>
    <img
      src="./public/img/null.jpg"
    />
    <button
      className="texture_remove"
      onClick={[Function]}
    >
      remove
    </button>
  </article>
</article>
`;

exports[`redo_undo: basicMaterial texture test simulate set currentSceneTreeNode test remove texture test undo operate test undo one step step which from first to zero 1`] = `
<article
  className="wonder-inspector-material"
>
  <article
    className="wonder-material-color"
  >
    <div
      className=""
    >
      <span
        className=""
      >
        color : 
      </span>
      <span
        className=""
      >
        #ffffff
      </span>
      <button
        className=""
        onClick={[Function]}
      >
        pick color
      </button>
    </div>
  </article>
  <article
    className="wonder-material-texture"
  >
    <div
      className="texture_ground"
      onDragEnter={[Function]}
      onDragLeave={[Function]}
      onDragOver={[Function]}
      onDrop={[Function]}
      style={
        Object {
          "opacity": "1",
        }
      }
    />
    <span
      className=""
    >
      texture:
    </span>
    <img
      src="texture5img"
    />
    <button
      className="texture_remove"
      onClick={[Function]}
    >
      remove
    </button>
  </article>
</article>
`;
